### PR TITLE
Sm/code-for-gacks-and-typeErrors

### PR DESCRIPTION
**Pull request by mshanemc**
detect gacks and typeErrors, assing them an exit code

some other cleanup
- current dev-scripts
- remove ts-sinon
- functional alternative to array mutations
- deprecate the protected SfCommand.assignProject so we can remove it from the API in the future 

@W-14408677@
@W-14408683@

### CODE REVIEW

File: src/errorHandling.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  if (errorIsTypeError(e)) {
+    return 10;
+  }
+
+  if ('exitCode' in e && typeof e.exitCode === 'number') {
+    return e.exitCode;
+  }
+
+  if ('code' in e) {
```
Comment (mshanemc) [inline src/errorHandling.ts:37]: `code` is typed quite awkwardly in SfError

File: src/errorHandling.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    return typeof e.code !== 'number' ? 1 : e.code;
+  }
+
+  return process.exitCode ?? 1;
+};
+
+/** identifies gacks via regex.  Searches the error message, stack, and recursively checks the cause chain */
+export const errorIsGack = (error: Error | SfError): boolean => {
+  /** see test for samples  */
+  const gackRegex = /\d{9,}-\d{3,} \(-?\d{7,}\)/;
```
Comment (mshanemc) [inline src/errorHandling.ts:47]: I just grabbed a LOT of gacks to see what the patterns looked like.

File: src/sfCommand.ts [hunk 3]
```diff
@@ -476,49 +472,6 @@ export abstract class SfCommand<T> extends Command {
...
 
-  /**
-   * Format errors and actions for human consumption. Adds 'Error (<ErrorCode>):',
-   * When there are actions, we add 'Try this:' in blue
-   * followed by each action in red on its own line.
-   * If Error.code is present it is output last in parentheses
-   *
-   * @returns {string} Returns decorated messages.
-   */
-  protected formatError(error: SfCommand.Error): string {
```
Comment (shetzel) [inline src/sfCommand.ts:487]: Removing this method could break commands that extend SfCommand.

File: test/unit/errorHandling.test.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        typeError = e;
+      }
+    }
+  });
+  it('matches on TypeError as error name', () => {
+    expect(errorIsTypeError(typeError)).to.be.true;
+  });
+
+  it('matches on TypeError in ', () => {
+    expect(errorIsTypeError(typeError)).to.be.true;
```
Comment (shetzel) [inline test/unit/errorHandling.test.ts:31]: This is the same test as the one above it.

File: src/sfCommand.ts (hunk 3 above)
Comment (mshanemc) [reply src/sfCommand.ts:487]: we don't have any

https://github.com/search?q=%28org%3Asalesforcecli+OR+org%3Aforcedotcom%29+formatError+language%3ATypeScript++NOT+is%3Aarchived&type=code

I couldn't find any outside of our repos, either.

File: test/unit/errorHandling.test.ts (hunk 4 above)
Comment (mshanemc) [reply test/unit/errorHandling.test.ts]: I meant to come back and finish that.  Nice catch!

**Comment by shetzel** [comment]
QA:
Built this branch, linked it into plugin-settings, then modified the alias:list command to throw 1. TypeError , 2. Gack-looking error.  Results:
1.
```
{
  "code": 10,
  "context": "AliasList",
  "commandName": "AliasList",
  "message": "a type error!",
  "name": "TypeError",
  "status": 10,
  "stack": "TypeError: a type error!\n    at AliasList.run (file:///Users/shetzel/dev/salesforcecli/plugin-settings/lib/commands/alias/list.js:22:15)\n    at AliasList._run (/Users/shetzel/dev/salesforcecli/sf-plugins-core/node_modules/@oclif/core/lib/command.js:303:33)",
  "exitCode": 10,
  "warnings": []
}
2.
```
{
  "code": 20,
  "context": "AliasList",
  "commandName": "AliasList",
  "message": "UNKNOWN_EXCEPTION: An unexpected error occurred. Please include this ErrorId if you contact support: 859937875-93964 (963231933)",
  "name": "Error",
  "status": 20,
  "stack": "Error: UNKNOWN_EXCEPTION: An unexpected error occurred. Please include this ErrorId if you contact support: 859937875-93964 (963231933)\n    at AliasList.run (file:///Users/shetzel/dev/salesforcecli/plugin-settings/lib/commands/alias/list.js:22:15)\n    at AliasList._run (/Users/shetzel/dev/salesforcecli/sf-plugins-core/node_modules/@oclif/core/lib/command.js:303:33)",
  "exitCode": 20,
  "warnings": []
}
```

**Pull request merged**